gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/DatasetTest.java

```diff
@@ -49,219 +61,291 @@
   private static final ViewDefinition VIEW_DEFINITION = ViewDefinition.of("QUERY");
   private static final ExternalTableDefinition EXTERNAL_TABLE_DEFINITION =
       ExternalTableDefinition.of(ImmutableList.of("URI"), Schema.of(), FormatOptions.csv());
-  private static final Iterable<TableInfo> TABLE_INFO_RESULTS = ImmutableList.of(
-      TableInfo.builder(TableId.of("dataset", "table1"), TABLE_DEFINITION).build(),
-      TableInfo.builder(TableId.of("dataset", "table2"), VIEW_DEFINITION).build(),
-      TableInfo.builder(TableId.of("dataset", "table2"), EXTERNAL_TABLE_DEFINITION).build());
+  private static final TableInfo TABLE_INFO1 =
+      TableInfo.builder(TableId.of("dataset", "table1"), TABLE_DEFINITION).build();
+  private static final TableInfo TABLE_INFO2 =
+      TableInfo.builder(TableId.of("dataset", "table2"), VIEW_DEFINITION).build();
+  private static final TableInfo TABLE_INFO3 =
+      TableInfo.builder(TableId.of("dataset", "table3"), EXTERNAL_TABLE_DEFINITION).build();
 
-  @Rule
-  public ExpectedException thrown = ExpectedException.none();
+  private BigQuery serviceMockReturnsOptions = createStrictMock(BigQuery.class);
+  private BigQueryOptions mockOptions = createMock(BigQueryOptions.class);
   private BigQuery bigquery;
+  private Dataset expectedDataset;
   private Dataset dataset;
 
-  @Before
-  public void setUp() throws Exception {
+  private void initializeExpectedDataset(int optionsCalls) {
+    expect(serviceMockReturnsOptions.options()).andReturn(mockOptions).times(optionsCalls);
+    replay(serviceMockReturnsOptions);
     bigquery = createStrictMock(BigQuery.class);
-    dataset = new Dataset(bigquery, DATASET_INFO);
+    expectedDataset = new Dataset(serviceMockReturnsOptions, new Dataset.BuilderImpl(DATASET_INFO));
+  }
+
+  private void initializeDataset() {
+    dataset = new Dataset(bigquery, new Dataset.BuilderImpl(DATASET_INFO));
   }
 
   @After
   public void tearDown() throws Exception {
-    verify(bigquery);
+    verify(bigquery, serviceMockReturnsOptions);
   }
 
   @Test
-  public void testInfo() throws Exception {
-    assertEquals(DATASET_INFO, dataset.info());
+  public void testBuilder() {
```

Comment:
There's no such constructor in functional objects. We should rather test `fromPb`.